### PR TITLE
Implement graphql mutation node for Clone Analysis framework

### PR DESCRIPTION
**Pull request by susilnem**
Addresses
- https://github.com/the-deep/server/issues/1476

## Changes
- Add mutation for clone analysis framework
- Add Separate serializer for the cloning

Mention related users here if any.

## This PR doesn't introduce any:

- [x] temporary files, auto-generated files or secret keys
- [x] n+1 queries
- [x] flake8 issues
- [x] `print`
- [x] typos
- [x] unwanted comments

## This PR contains valid:

- [x] tests
- [x] permission checks (tests here too)
- [x] translations

### CODE REVIEW

File: apps/analysis_framework/serializers.py [hunk 1]
```diff
@@ -708,3 +708,36 @@ def create(self, validated_data):
...
+    title = serializers.CharField(required=True)
+    description = serializers.CharField(required=False)
+    project = serializers.IntegerField(required=False)
+
+    class Meta:
+        model = AnalysisFramework
+        fields = ('title', 'description', 'project')
+
+    def validate(self, validated_data):
+        project_id = validated_data.get('project')
```
Comment (Rup-Narayan-Rajbanshi) [inline apps/analysis_framework/serializers.py:723]:
```suggestion
        project_id = validated_data.get('project', None)
```

File: apps/analysis_framework/serializers.py [hunk 2]
```diff
@@ -708,3 +708,36 @@ def create(self, validated_data):
...
+        # Check if project exists and user has access to it
+        if project_id is not None:
```
Comment (Rup-Narayan-Rajbanshi) [inline apps/analysis_framework/serializers.py]:
```suggestion
        if project_id:
```

File: apps/analysis_framework/serializers.py (hunk 1 above)
Comment (Rup-Narayan-Rajbanshi) [inline apps/analysis_framework/serializers.py]: Can be made a bit simpler.

File: apps/analysis_framework/serializers.py (hunk 2 above)
Comment (Rup-Narayan-Rajbanshi) [reply apps/analysis_framework/serializers.py]: Can be made a bit simpler.

**Review by Rup-Narayan-Rajbanshi — COMMENTED** [review]
Looks good, just some minor code formatting.

**Review by Rup-Narayan-Rajbanshi — DISMISSED** [review]

File: apps/analysis_framework/serializers.py [hunk 3]
```diff
@@ -708,3 +708,36 @@ def create(self, validated_data):
...
+
+
+class AnalysisFrameworkCloneGlSerializer(serializers.Serializer):
+    title = serializers.CharField(required=True)
+    description = serializers.CharField(required=False)
+    project = serializers.IntegerField(required=False)
+
+    class Meta:
+        model = AnalysisFramework
+        fields = ('title', 'description', 'project')
```
Comment (thenav56) [inline apps/analysis_framework/serializers.py]: Do we need this for serializers.Serializer?

File: apps/analysis_framework/serializers.py [hunk 4]
```diff
@@ -708,3 +708,36 @@ def create(self, validated_data):
         validated_data['framework'] = self.framework
         validated_data['added_by'] = self.context['request'].user
         return super().create(validated_data)
+
+
+class AnalysisFrameworkCloneGlSerializer(serializers.Serializer):
```
Comment (thenav56) [inline apps/analysis_framework/serializers.py]: We don't need `Gl` here, Let's use AnalysisFrameworkCloneSerializer
> NOTE: We used to add ...GqSer.. when we already had serializer for REST.

File: apps/analysis_framework/serializers.py [hunk 5]
```diff
@@ -708,3 +708,36 @@ def create(self, validated_data):
...
+    title = serializers.CharField(required=True)
+    description = serializers.CharField(required=False)
+    project = serializers.IntegerField(required=False)
```
Comment (thenav56) [inline apps/analysis_framework/serializers.py]: Let's use Primary ID field here.

File: apps/analysis_framework/serializers.py [hunk 6]
```diff
@@ -708,3 +708,36 @@ def create(self, validated_data):
...
+        fields = ('title', 'description', 'project')
+
+    def validate(self, validated_data):
+        project_id = validated_data.get('project')
+        # Check if project exists and user has access to it
+        if project_id is not None:
+            project = Project.objects.filter(id=project_id).first()
+            if project is None:
+                raise serializers.ValidationError('Invalid project ID')
+            if not project.can_modify(self.context['request'].user):
```
Comment (thenav56) [inline apps/analysis_framework/serializers.py]: We have can_clone, let's use that for now

File: apps/analysis_framework/serializers.py [hunk 7]
```diff
@@ -708,3 +708,36 @@ def create(self, validated_data):
...
+
+    def create(self, validated_data):
+        af = self.context['request'].active_af
+        new_af = af.clone(self.context['request'].user, validated_data)
+        new_af.add_member(self.context['request'].user, new_af.get_or_create_owner_role())
+
+        if project := validated_data.get('project'):
+            project.analysis_framework = new_af
+            project.modified_by = self.context['request'].user
+            project.save(update_fields=['analysis_framework', 'modified_by'])
```
Comment (thenav56) [inline apps/analysis_framework/serializers.py]: Can you confirm if modified_at also updates?

File: apps/analysis_framework/tests/test_mutations.py [hunk 8]
```diff
@@ -774,6 +774,73 @@ def _query_check(**kwargs):
...
+                    title
+                    description
+                    clonedFrom
+                  }
+                }
+              }
+            }
+        '''
+
+        member_user = UserFactory.create()
```
Comment (thenav56) [inline apps/analysis_framework/tests/test_mutations.py:804]: We will need additional users here.

Basically,
- User with normal access to project  & AF <- Fail
- User with normal access to project & but not AF (Private) <- Fail

- User with admin access to project but not AF (Private) <- Fail
- User with admin access to project & AF <- Success
- User with admin access to project & AF (Private with access) <- Success


Let's add check that we don't create any additional AF or change project.af until it is set.

File: apps/analysis_framework/tests/test_mutations.py [hunk 9]
```diff
@@ -774,6 +774,73 @@ def _query_check(**kwargs):
...
+        _query_check(assert_for_error=True)
+        # ---------- With login
+        self.force_login(non_member_user)
+        # ---------- Let's Clone a new AF (Using create test data)
+        _query_check(assert_for_error=True)
+
+        # ---------- With login (with access member)
+        self.force_login(member_user)
+        response = _query_check()
+        self.assertMatchSnapshot(response, 'success')
```
Comment (thenav56) [inline apps/analysis_framework/tests/test_mutations.py]: Let's replace snapshot test with normal dict check.

**Review by thenav56 — CHANGES_REQUESTED** [review]
Looks good. Few changes required

File: apps/analysis_framework/serializers.py [hunk 10]
```diff
@@ -708,3 +708,25 @@ def create(self, validated_data):
...
+    description = serializers.CharField(required=False)
+    project = serializers.PrimaryKeyRelatedField(queryset=Project.objects.all(), required=False)
+
+    def validate_project(self, project):
+        if not project.can_modify(self.context['request'].user):
+            raise serializers.ValidationError('User does not have permission to modify the project')
+        return project
+
+    def create(self, validated_data):
+        af = self.context['request'].active_af
```
Comment (thenav56) [inline apps/analysis_framework/serializers.py]: We need this check as well
https://github.com/the-deep/server/blob/e951e7892d1618a28d02d92b544f5e788980d7d2/apps/analysis_framework/views.py#L100-L101

**Review by thenav56 — CHANGES_REQUESTED** [review]

File: apps/analysis_framework/tests/test_mutations.py (hunk 8 above)
Comment (susilnem) [reply apps/analysis_framework/tests/test_mutations.py:804]: @AdityaKhatri  For
> User with admin access to project & AF (Private with access) <- Success

Do we need to check the clone AF is also private?

File: apps/analysis_framework/serializers.py (hunk 7 above)
Comment (susilnem) [reply apps/analysis_framework/serializers.py]: Yes, Modified_by field is working fine.

Comment (thenav56) [reply apps/analysis_framework/serializers.py]: Check if `modified_at` is also updated. Or let's just add that field to the update_fields for visibility

Comment (susilnem) [reply apps/analysis_framework/serializers.py]: Checked! `Modified_at` is also updated.

**Review by thenav56 — APPROVED** [review]

File: apps/analysis_framework/serializers.py [hunk 11]
```diff
@@ -711,34 +711,33 @@ def create(self, validated_data):
...
-            validated_data['project'] = project
-        return validated_data
+    def validate_af_id(self, af_id):
+        try:
+            af = AnalysisFramework.get_for_gq(self.context['request'].user, only_member=False).get(pk=af_id)
+        except AnalysisFramework.DoesNotExist:
+            raise serializers.ValidationError('Analysis Framework does not exist')
+        if not af.can_clone(self.context['request'].user):
+            raise serializers.ValidationError('User does not have permission to clone the Analysis Framework')
+        return af
```
Comment (thenav56) [inline apps/analysis_framework/serializers.py]: Let's use if/else 
```suggestion
    def validate_af_id(self, af_id):
        af = (
            AnalysisFramework.
            get_for_gq(self.context['request'].user, only_member=False)
            .filter(pk=af_id)
            .first()
        )
         if af is None:
            raise serializers.ValidationError('Analysis Framework does not exist')
         if not af.can_clone(self.context['request'].user):
            raise serializers.ValidationError(
                'User does not have permission to clone the Analysis Framework'
            )
        return af
```

File: apps/analysis_framework/serializers.py [hunk 12]
```diff
@@ -711,34 +711,33 @@ def create(self, validated_data):
...
+
+    def validate_project(self, project):
+        if not project.can_modify(self.context['request'].user):
+            raise serializers.ValidationError('User does not have permission to modify the project')
+        return project
 
     def create(self, validated_data):
-        af = self.context['request'].active_af
+        af = validated_data['af_id']
```
Comment (thenav56) [inline apps/analysis_framework/serializers.py:742]:
```suggestion
        # NOTE: af is an obj, check validate_af_id
        af = validated_data['af_id']
```

File: apps/analysis_framework/serializers.py [hunk 13]
```diff
@@ -709,3 +709,35 @@ def create(self, validated_data):
...
+        return af
+
+    def validate_project(self, project):
+        if not project.can_modify(self.context['request'].user):
+            raise serializers.ValidationError('User does not have permission to modify the project')
+        return project
+
+    def create(self, validated_data):
+        af = validated_data['af_id']
+        new_af = af.clone(self.context['request'].user, validated_data)
```
Comment (thenav56) [inline apps/analysis_framework/serializers.py]: Let's also refactor the existing clone logic, replacing the dict to arguments

```diff
diff --git a/apps/analysis_framework/models.py b/apps/analysis_framework/models.py
index a3217814b..6b45f80fe 100644
--- a/apps/analysis_framework/models.py
+++ b/apps/analysis_framework/models.py
@@ -1,5 +1,5 @@
 import copy
-from typing import Union
+from typing import Union, Optional
 
 from django.core.exceptions import ValidationError
 from django.db import models
@@ -63,13 +63,14 @@ class AnalysisFramework(UserResource):
     def __str__(self):
         return self.title
 
-    def clone(self, user, overrides={}):
+    def clone(self, user, title: Optional[str] = None, description: Optional[str] = None):
         """
         Clone analysis framework along with all widgets,
         filters and exportables
         """
-        title = overrides.get("title", "{} (cloned)".format(self.title[:230]))  # Strip off extra chars from title
-        description = overrides.get("description", "")
+        title = title or "{} (cloned)".format(self.title[:230])
+        description = description or ""
+
         clone_analysis_framework = AnalysisFramework(
             title=title,
             description=description,
```

**Review by thenav56 — CHANGES_REQUESTED** [review]
Let's do some clean-ups

**Review by sudan45 — APPROVED** [review]

**Comment by sudan45** [comment]
Looks Good

File: apps/analysis_framework/serializers.py [hunk 14]
```diff
@@ -709,3 +709,42 @@ def create(self, validated_data):
...
+
+    def validate_project(self, project):
+        if not project.can_modify(self.context['request'].user):
+            raise serializers.ValidationError('User does not have permission to modify the project')
+        return project
+
+    def create(self, validated_data):
+        # NOTE: af is an obj, check validate_af_id
+        af = validated_data['af_id']
+        new_af = af.clone(self.context['request'].user, validated_data['title'], validated_data.get('description'))
```
Comment (thenav56) [inline apps/analysis_framework/serializers.py]: Let's use keyword for keyword arguments
```suggestion
        new_af = af.clone(self.context['request'].user, title=validated_data['title'], description=validated_data.get('description'))
```

**Review by thenav56 — CHANGES_REQUESTED** [review]
Almost done.

**Review by thenav56 — APPROVED** [review]

**Pull request merged**